### PR TITLE
[Attention] Remove DCP indexer interleave guard and test TP1 output parity

### PR DESCRIPTION
**Pull request by stu-cao**
## Purpose

Remove the sparse indexer's rejection of DCP with `cp_kv_cache_interleave_size > 1`, and add direct attention-output regression tests against complete, unsharded TP1 KV. This enables review of configurations such as DCP with NIXL, which uses interleave 64.

**Draft: the tested attention outputs agree within the stated FP8 tolerances. Full-model outputs are numerically sensitive, and accuracy equivalence is not established.** The original single GSM8K pair scored 11 questions lower at interleave64. Three further controlled pairs scored **23 higher, 3 higher, and 10 lower**; the lower score is not consistent across repeats. The investigation below causally identifies attention rounding and index-order variability as sources of changed generations.

The production change is exactly the nine-line guard deletion in `DeepseekV32IndexerMetadataBuilder`. It is generic because this builder implements interleave/ownership arithmetic independently of model identity or KV-transfer connector. There is no GLM/NIXL allowlist or backend capability override. Separate backend and MTP compatibility checks remain unchanged. The evidence is limited to the tested implementations/configurations, not every configuration admitted by deleting the guard.

The original guard in #46076 (`c5200d3565`) cited GSM8K failure. **That historical failure has not been root-caused.** The pure-DCP defects reported in #50095 and fixed by #50005, and the non-owner-rank prefill fix in #54908, are plausible contributors; both fixes are in this branch's base. Neither establishes why interleave1 passed while larger interleaves failed. The historical result used GLM-5.2-FP8/lm_eval at concurrency128; the current GLM-5.3-NVFP4 evaluation is not a direct reproduction.

Duplicate-work searches for DCP/interleave and #50095 found no other open PR combining this guard deletion with TP1 attention-output comparisons. #50005/#54908 are merged; #54541 is a backport and #52472 adds a different backend.

## Test Plan

Tests are committed to the existing vLLM attention suites. Each comparison uses identical generated inputs: TP1/DCP1 has complete KV; TP4/DCP4 shards the same KV with interleave64 and combines actual distributed attention outputs. The real indexer builder is constructed from the DCP configuration and `build()` is called. Restoring the guard fails its constructor before attention executes.

- `test_sparse_dcp4_interleave64_attention_matches_tp1[flashinfer,flashmla]` exercises real indexer quantization, logits, local/global sparse selection, prefill/decode metadata, query gathering, attention kernels and distributed output merging. `FLASHINFER_MLA_SPARSE` resolves to the TRTLLM sparse implementation with FP8 Q/KV and A2A; `FLASHMLA_SPARSE` uses BF16 Q, `fp8_ds_mla` and `ag_rs`.
- `test_trtllm_ragged_dcp4_interleave64_attention_matches_tp1` exercises the internal `TRTLLM_RAGGED` dense MLA prefill implementation through production chunk metadata and `forward_mha`: FP8 cache gathering/dequantization, KV projection, causal suffix/context attention, continuation chunks and combined outputs.

Nine cases per sparse backend cover two seeds, decode and four causal query rows (MTP3), two scale/input-magnitude settings, top-k2048, contexts through8193, 64/256-token boundaries, shuffled pages, populated future tokens and empty shards. The ordinary sparse-prefill case uses64 query tokens and lengths65/8193. Its position keys and one-hot indexer queries give exact, strictly ordered scores, allowing independent assertions of the expected causal top-k. Eight dense-prefill cases cover two seeds, suffix lengths4/65, scales1/0.25, contexts0/1/63/64/65/255/256/257/2049 and multiple context chunks. Tested geometry is64 heads, KV rank512, NoPE192, RoPE64, and dense value dimension256.

The tests retain actual TP1/TP4 outputs, input hashes, quantized query bytes, indices and causal bounds in pytest's temporary directory. They assert exact input/index/bound equality. The FP32 oracle validates attention given the real indexer's selections and catches shared attention-kernel errors; it cannot detect a top-k error shared by both paths. The independently known prefill ranking checks selection separately.

Sparse tolerances follow the existing backend budget: elementwise `rtol=0.065, atol=0.05`, plus per-query relative L2 below0.065 against TP1 and the oracle. Dense prefill uses `rtol=atol=0.01` and per-query relative L2 below0.01. Sparse natural-log LSE uses `rtol=atol=1e-4`. These are numerical tolerances, not bitwise equality.

```bash
export VLLM_FLASHINFER_WORKSPACE_BUFFER_SIZE=2147483648
export NCCL_IB_DISABLE=1 NCCL_SOCKET_IFNAME=lo GLOO_SOCKET_IFNAME=lo
export OMP_NUM_THREADS=8
PYTHONPATH=/work/source /work/.venv/bin/python -m pytest \
  /work/source/tests/v1/attention/test_indexer_dcp_localize.py \
  /work/source/tests/v1/attention/test_mla_backends.py::test_trtllm_ragged_dcp4_interleave64_attention_matches_tp1 \
  --basetemp=/work/pytest-prefill-final \
  --junitxml=/work/pytest-prefill-final.xml -q -s
```

## Test Result

**79 tests passed in109.02s with standard parent conftest/autouse fixtures, without `--confcutdir`, on four GB200 GPUs.** This includes26 actual output comparisons with the guard completely removed.

| Implementation | Cases | Max absolute difference | Worst query relative L2 | Index-set mismatches |
|---|---:|---:|---:|---:|
| `FLASHINFER_MLA_SPARSE` / TRTLLM |9|0.03125|1.93911%|0|
| `FLASHMLA_SPARSE` |9|0.0078125|0.45789%|0|
| `TRTLLM_RAGGED` |8|0.00390625|0.23980%|N/A|

Maximum sparse natural-log LSE difference was `9.536743e-7`, including comparison with the FP32 oracle. The tests populate caches directly; fused model-level cache population is exercised separately by the model investigation. Kernel MTP cases do not validate the full MTP scheduler or relax its separate compatibility checks.

Runtime: `vllm/vllm-openai@sha256:275cf4900543d71daaf43b008619c79372e139c0e4f4b643202b3fec97886c19`, torch2.13.0+cu130, CUDA13.0, FlashInfer0.6.18, CUTLASS DSL4.6.2, Triton3.7.1. Checkout Python source is `9005dcfd50`; native artifacts are from `d9105ea800`, with native source unchanged between those commits. Production source is unchanged from the original model run at `ce03ef5450`. Actual numerical-test tensors are retained in a verified archive, SHA256 `50edbd7dbc1262d945d681bdc2879a8e5b95b6b228c938100d71d95a9c8e56d5`.

Applicable manual-stage pre-commit hooks, including Python3.10–3.13 mypy, passed locally for the changed files; `git diff --check` passed. GitHub's contributor-eligibility pre-check blocks its pre-commit job; no eligibility/workflow bypass is included. The standard-conftest run does not reproduce every Buildkite environment setting. No DCP-specific end-to-end CI step is added.

### Model evaluation and root-cause investigation

All model runs are **aggregated TP4/DCP4 serving**, with `Inferact/GLM-5.3-NVFP4` revision `ce67b36f3669192b5bb233819f0fda6c8a9837f8`, the configured FlashInfer sparse backend/internal TRTLLM prefill, FP8 KV, block64, prefix caching, max length16384, max batched tokens8192 and max sequences32. MTP and P/D/NIXL transport are not configured. The unchanged five-shot GSM8K prompt set contains all1319 questions; scoring uses this checkout's GSM8K helper, greedy temperature0, seed0 and max1024 completion tokens.

| Run | Interleave1 correct /1319 | Interleave64 correct /1319 | I64 minus I1 |
|---|---:|---:|---:|
| Original pair |1206 (91.43%)|1195 (90.60%)|−11|
| Controlled repeat0 |1194 (90.52%)|1217 (92.27%)|+23|
| Controlled repeat1 |1205 (91.36%)|1208 (91.58%)|+3|
| Controlled repeat2 |1214 (92.04%)|1204 (91.28%)|−10|

The controlled repeats use identical integer prompts, graph execution, concurrency32, a fresh shared-prefix cache salt for each repeat, and the same saved MoE kernel tactics on both hosts. Independently selected tactics initially differed in10/21 entries; pinning removes that confound within these repeats. Batching remains asynchronous. Hosts were not swapped. I1 varied by20 correct answers and I64 by13 across repeats; only129–151/1319 full sequences matched in within-configuration comparisons. Mean I64-minus-I1 accuracy was+0.404 percentage points. A descriptive two-level bootstrap over these three runs and shared questions gives a95% percentile interval of[−1.137,+1.996] points. **This does not prove accuracy equivalence or uniquely explain the original11-question decrease.**

The original pair had53 correct→wrong and42 wrong→correct, with no request errors, invalid answers or truncations. The new pairs had31/54,45/48 and46/36 flips respectively. All7914 requests completed with no invalid parsed answers; one I1 response in repeat1 reached the token limit while repeating an incorrect calculation and scored wrong. All raw responses and scores are retained locally in a verified archive, SHA256 `c0c4470b0e5ee1a16c5dd5be49b56eec7c9c701deef7c511cec147167be692b2`.

The layer traces and causal interventions use eager serving, one in-flight request per server, fresh per-request cache salts, and recorded token IDs/top20 logprobs. They separate selection/addressing from arithmetic:

1. The DCP top-k merge uses stable score/token-ID membership keys but an atomic counter to assign output positions. Repeated cold eager generations matched only1/8 within each interleave; diagnostic canonical index ordering made8/8 full sequences and returned logprobs exact within each configuration. Sorting did not establish equality between interleaves.
2. On a traced first decode, layer0 Q/KV bytes were identical between interleaves and selected IDs were exactly all causal tokens. Native combined outputs differed by0.39761% relative L2. Query/current-token KV differences began at layer1; all prompt KV bytes stayed identical through78 layers. The model difference reached27.67625% at layer77 after its inputs had diverged.
3. A fixed numerical model of128-key tiles, FP8 softmax-probability rounding and BF16 unnormalized partial rounding explains98.96% of the squared first-layer interleave difference. It matches about99.75% of elements exactly across448 separate saved rank-local GPU outputs, with worst residual below0.057%. This isolates a numerical mechanism; it is not an instruction-level kernel reconstruction.
4. With common MoE tactics, rank-local FP64 attention and FP64 DCP combination matched complete-KV FP64 at all78 traced combined outputs and on both selected full-answer cases (13/15). Explicit synchronization and a discarded complete-KV gather did not remove native divergence. Changing only the merge to FP32 also did not remove it. Complete-KV FP8 likewise matched across interleaves, but produced different answers from FP64. Agreement is distinct from answer correctness.
5. Rank-local FP32 attention/merge differed in only2/32768 layer0 BF16 values near rounding midpoints, yet its model generations diverged. An additional controlled perturbation of one layer0 BF16 value to its adjacent representable value changed subsequent generated wording while every attention computation used FP64; the input hashes matched, and both interleaves agreed under each perturbation. The parsed answer stayed unchanged in that isolated test.

A BF16-kernel intervention preserved the existing FP8-grid Q/KV values, selected physical rows, scales and DCP combination. It reduced same-input interleave output difference from0.397685% to0.261527% in a GPU replay, but did not remove serving divergence on cases13/15; both modes were exact within each configuration across two cold repeats. Case15 changed from I1/I64 answers125/2971 with native FP8 to29/125 with BF16 (label125). This selected subset is not an accuracy estimate or a proposed production fix.

All original GSM8K sequences were below top-k2048, so equal-score cutoff ambiguity above top-k cannot explain that corpus. The failed initial random prefill fixture had equal cutoff scores and was replaced with the independently ranked fixture described above. No causal/KV-addressing defect was found in the captured cases. Conditional checks against FP64 attention on each layer's own captured inputs gave maximum whole-output errors of0.9553% (I1) and2.4428% (I64); these do not assert equality of later model activations.

Actual combined outputs for all78 layers across the five traced modes/two interleaves, Q/KV bytes, indices and representative partials/LSE are retained in a verified archive, SHA256 `272db45212eda347596ce069ced856b432dbca2ca923a7bb2aa01b75050e8206`. Diagnostic interventions are isolated runtime code, not production changes in this PR. Raw artifacts are retained locally and are not attached here.

Graph-enabled server command (set `CP_INTERLEAVE` to1 or64, with the environment variables above and `PYTHONHASHSEED=0`):

```bash
PYTHONPATH=/work/source /work/.venv/bin/python -m vllm.entrypoints.cli.main serve \
  /models/models--Inferact--GLM-5.3-NVFP4/snapshots/ce67b36f3669192b5bb233819f0fda6c8a9837f8 \
  --served-model-name dcp-gsm8k-review --host 0.0.0.0 --port 8000 \
  --tensor-parallel-size 4 --decode-context-parallel-size 4 \
  --cp-kv-cache-interleave-size "$CP_INTERLEAVE" \
  --attention-backend FLASHINFER_MLA_SPARSE --kv-cache-dtype fp8 \
  --block-size 64 --enable-prefix-caching --max-model-len 16384 \
  --max-num-batched-tokens 8192 --max-num-seqs 32 \
  --gpu-memory-utilization 0.87 --distributed-executor-backend mp \
  --seed 0 --no-enable-log-requests
```

For the original evaluation protocol against either graph-enabled server:

```bash
PYTHONPATH=/work/source /work/.venv/bin/python \
  /work/source/tests/evals/gsm8k/gsm8k_eval.py \
  --host http://127.0.0.1 --port 8000 --num-questions 1319 --num-shots 5 \
  --max-tokens 1024 --temperature 0 --seed 0 --max-concurrency 32 \
  --save-results /work/gsm8k-summary.json
```

The repeated-run wrapper additionally freezes integer prompt IDs, asserts returned prompt IDs, sets per-repeat cache salts, pins kernel tactics, and saves every response. The helper command alone does not implement those additional controls.

## AI assistance

OpenAI Codex assisted with code, GPU tests, investigation and PR preparation under the submitter's direction. Commits include attribution and sign-off. Human review of every changed line and the generic guard removal remains outstanding; this draft does not claim that the submitter personally reran the automated validation.

### CODE REVIEW

**Review by GirasoleY — COMMENTED** [review]
Thank you for helping us adding the E2E tests! 

I opened one change to add a 4 node B200 test setup as the current test gets skipped with < 4GPU nodes. feel free to merge this in: https://github.com/stu-cao/vllm/pull/1/changes

From the test results we can relax the guard. However your description mentioned a gsm8k score drop, could you confirm if it is an actual regression or just red herring?

Once both are addressed we are good to go.

**Comment by stu-cao** [comment]
@GirasoleY The GSM8K discrepancy seems to be a re-herring. I ran the test 4 times. interleave 64 has higher score twice while interleave 1 has higher score twice.
```

Run | Interleave1 correct /1319 | Interleave64 correct /1319 | I64 minus I1
-- | -- | -- | --
Original pair | 1206 (91.43%) | 1195 (90.60%) | −11
Controlled repeat0 | 1194 (90.52%) | 1217 (92.27%) | +23
Controlled repeat1 | 1205 (91.36%) | 1208 (91.58%) | +3
Controlled repeat2 | 1214 (92.04%) | 1204 (91.28%) | −10
```

**Review by LucasWilkinson — APPROVED** [review]
LGTM assuming @GirasoleY is happy with it now

**Review by mgoin — APPROVED** [review]
